[ashleyang2001] iP

ashleyang2001 opened the pull request.

ashleyang2001 committed:
Renamed Duke class to Prune.

Implemented greeting and exit.

ashleyang2001 committed:
Modified main method to echo commands entered by the user, and exits program when the user types the command "bye".

ashleyang2001 committed:
Added method addTasks(String Task) to store text entered by user.

Added method printTasks() to display all tasks entered by user when requested using command  "list".

ashleyang2001 committed:
Adjusted code to comply with given coding standard.

ashleyang2001 committed:
Added the ability to mark tasks as done or not done using markTask(String input).

Added class Task to encapsulate functionality of a task.

(cherry picked from commit 1cffefbfd26873836546ea41b69b56280ac26413)

ashleyang2001 committed:
Merge commit '50f1d27548c78730b1b10c078dee0ae94c49bda2'

ashleyang2001 committed:
Added support for tracking 3 types of tasks: ToDo, Deadline, Event extending Task Class.

Encapsulate interaction with user in UserInterface Class and input parsing of input in Parser for better abstraction.

ashleyang2001 committed:
Modified runtest.bat to work with current directory to perform automated text UI testing.

ashleyang2001 committed:
Gave meaningful name to temporary variables.

jwyk reviewed (commented):
Overall, LGTM. I like the structure of the project and all the classes seem good to me too. In the future, it might be possible to investigate a better way of processing inputs other than an array, but your solution works well.

jwyk commented on src/main/java/Prune.java:
```diff
@@ -0,0 +1,68 @@
+public class Prune {
+
+    public static final int MAX_NUMBER_OF_TASKS = 100;
+    private final Task[] tasks = new Task[MAX_NUMBER_OF_TASKS];
```
in the future, could probably replace this with an ArrayList? It's easier to maintain compared to an array.

jwyk reviewed (commented):
Refer to above for overall comments

denzelloke commented on src/main/java/UserInterface.java:
```diff
@@ -0,0 +1,48 @@
+     */
+    public static void greet() {
+        System.out.println("\tHello! I am Prune.");
+        String mascot = """
```
Cute! good touch of personalisation

denzelloke commented on src/main/java/Parser.java:
```diff
@@ -0,0 +1,94 @@
+        } else if (command.equals(MARK_DONE_COMMAND) || command.equals(MARK_NOT_DONE_COMMAND)) {
+            // Process mark/unmark commands
+            try {
+                int taskIndex = Integer.parseInt(inputArray[1]) - 1;
```
i think this - 1 might be a case of magic number

denzelloke reviewed (commented):
sound application of SWE principles, no further comments from me

ashleyang2001 committed:
Handled error using InvalidCommand and InvalidTaskNumber exceptions, added try catch to catch invalid task description formatting.

ashleyang2001 committed:
Merge branch 'branch-Level-5'

ashleyang2001 committed:
Organised classes into Prune, Exceptions, Tasks packages.

ashleyang2001 committed:
Changed Task array type to ArrayList, added support for deleting tasks from list.

ashleyang2001 committed:
Added saving of most updated task list into saved_tasks.txt and ability to load the data on start up.

ashleyang2001 committed:
Removed write funcs comments in Parser.

ashleyang2001 committed:
Create saved_tasks.txt in folder data if file does not exist, relative to root.

ashleyang2001 committed:
Extracted out classes Ui, Storage, TaskList. Same features as the previous commit.

ashleyang2001 committed:
Fixed saving tasks after every command.

ashleyang2001 committed:
Added finding of matching tasks by searching for a keyword in the task description.

ashleyang2001 committed:
Added JavaDoc header comments to classes and methods

ashleyang2001 committed:
Merge pull request #1 from ashleyang2001/branch-Level-9

Added finding of matching tasks by searching for a keyword in the tas…

ashleyang2001 committed:
Merge branch 'branch-A-JavaDoc'

# Conflicts:
#	src/main/java/Prune/Parser.java

ashleyang2001 committed:
Merge pull request #2 from ashleyang2001/branch-A-JavaDoc

Added JavaDoc header comments to classes and methods

ashleyang2001 committed:
Modified dog mascot string.

ashleyang2001 committed:
Add files via upload

Product screenshot

okkhoy commented on src/main/java/Prune/Exceptions/InvalidCommand.java:
```diff
@@ -0,0 +1,17 @@
+package Prune.Exceptions;
```
package name should not be capitalized. better `prune.exception`

okkhoy commented on src/main/java/Prune/Parser.java:
```diff
@@ -0,0 +1,255 @@
+        case TODO_COMMAND:
+        case DEADLINE_COMMAND:
+        case EVENT_COMMAND:
+            task = processAdding(inputArray);
```
better name would be `processAddCommand` rather than `processAdding`. 
similarly other methods.

okkhoy commented on src/main/java/Prune/Parser.java:
```diff
@@ -0,0 +1,255 @@
+        case DELETE_COMMAND:
+            task = processDeleting(inputArray);
+            printDeleteTaskMsg(task);
+            break;
+        case FIND_COMMAND:
```
inconsistent vertical breaks. earlier you have a blank line between 2 case blocks

okkhoy commented on src/main/java/Prune/Parser.java:
```diff
@@ -0,0 +1,255 @@
+            throw new InvalidCommand(String.format("\tUnknown command: %s\n\tPlease try again", command));
+        }
+
+        // Write data to file for mark, todo, deadline, event, delete commands
```
this looks like a 'note-to-self' comment. should be removed.

okkhoy commented on src/main/java/Prune/Parser.java:
```diff
@@ -0,0 +1,255 @@
+     * @param input Input string containing the task data.
+     */
+    public void processSavedData(String input) {
+        String[] inputArray = input.split(" ", 3);
+        String markStatus = inputArray[0];
+        boolean isDone = markStatus.equals(MARK_DONE_COMMAND);
+        // Form add task input array by removing mark/unmark command
+        String[] newInputArray = Arrays.copyOfRange(inputArray, 1, inputArray.length);
```
I see a couple of low-level string manipulation code here - potentially can extract them to improve SLAP

okkhoy commented on src/main/java/Prune/Parser.java:
```diff
@@ -0,0 +1,255 @@
+            String details = input[1];
+            String[] inputArray;
+            String description;
+            Task task;
+            if (command.equals(TODO_COMMAND)) {
```
can add a blank line in between for separation

okkhoy commented on src/main/java/Prune/Parser.java:
```diff
@@ -0,0 +1,255 @@
+                task = new Deadline(description, by);
+            } else {
+                // Process Event
+                inputArray = details.split(" /from ");
+                description = inputArray[0];
```
can extract these kind of string manipulation to separate methods for better SLAP

okkhoy commented on src/main/java/Prune/Parser.java:
```diff
@@ -0,0 +1,255 @@
+            this.tasks.addTask(task);
+            return task;
+        } catch (Exception e) {
+            System.out.println(String.format("\tInsufficient descriptions given for command: %s\n\t" + "Please ensure the following format:\n" + "\t\ttodo description\n" + "\t\tdeadline description /by when\n" + "\t\tevent description /from start /to end", command));
```
long line alert - remember what is the limit for the number of characters in a line?

denzelloke replied:
this line q long, maybe something like this could be more readable?

System.out.printf("""
    \tInsufficient descriptions given for command: %s
    \tPlease ensure the following format:
    \t\t- todo description
    \t\t- deadline description /by when
    \t\t- event description /from start /to end
""", command);

(the formatting shows weirdly on comments but ukwim)

okkhoy commented on src/main/java/Prune/Storage.java:
```diff
@@ -0,0 +1,82 @@
+     * @param parser The parser used to process the data from the file.
+     */
+    public void load(Parser parser) {
+        // Create file if it does not exist
```
good, but can you do this as a private/helper method instead of in load?

okkhoy commented on src/main/java/Prune/Storage.java:
```diff
@@ -0,0 +1,82 @@
+     *
+     * @param tasks TaskList containing the tasks to write to the file.
+     */
+    public void writeData(TaskList tasks) {
```
strange! load will check and create the file but write doesn't do that?

okkhoy commented on src/main/java/Prune/Tasks/Deadline.java:
```diff
@@ -0,0 +1,43 @@
+ */
+public class Deadline extends Task {
+
+    protected String by; // The deadline date for the task
```
comment is redundant

okkhoy commented on src/main/java/Prune/Tasks/Task.java:
```diff
@@ -0,0 +1,85 @@
+ * Subclasses of Task should implement the abstract method to provide specific task functionality.
+ */
+public abstract class Task {
+    protected String description; // The description of the task
+    protected boolean isDone; // The completion status of the task
```
remove such redundant comments

okkhoy reviewed (commented):
@ashleyang2001 in general, good code 👍 
there are some minor issues, please look at the comments that follow.

ashleyang2001 committed:
Modified Prune User Guide.

Modified TaskList to allow for case-insensitivity matches using find method.

ashleyang2001 committed:
Added Editing the data file section in README.md

denzelloke commented on src/main/java/Prune/Parser.java:
```diff
@@ -0,0 +1,255 @@
+        boolean isDone = markStatus.equals(MARK_DONE_COMMAND);
+        // Form add task input array by removing mark/unmark command
+        String[] newInputArray = Arrays.copyOfRange(inputArray, 1, inputArray.length);
+        processAdding(newInputArray);
```
processAdding returns a Task type object, but over here its not assigned to anything.
could this be by design?

denzelloke commented on src/main/java/Prune/Parser.java:
```diff
@@ -0,0 +1,255 @@
+    public Task processMarking(String[] input) throws InvalidTaskNumber {
+        String command = input[0];
+        try {
+            int taskIndex = Integer.parseInt(input[1]) - 1;
```
wait im not v sure BUT
if input[1] is null then you might have a nullpointerexception
perhaps u cd try this in your testing

denzelloke commented on src/main/java/Prune/Parser.java:
```diff
@@ -0,0 +1,255 @@
+     */
+    public Task processDeleting(String[] input) throws InvalidTaskNumber {
+        try {
+            int taskIndex = Integer.parseInt(input[1]) - 1;
```
same thing here, can check if invalidtasknumber will catch a nullpointerexception

denzelloke commented on src/main/java/Prune/Parser.java:
```diff
@@ -0,0 +1,255 @@
+    }
+
+    public TaskList processFinding(String[] input) {
+        String keyword = input[1];
```
could double check what happens in the exception that user only types "find" and nothing else

denzelloke commented on src/main/java/Prune/Prune.java:
```diff
@@ -0,0 +1,57 @@
+        Prune chatbot = new Prune("./data/saved_tasks.txt");
+        chatbot.run();
+    }
+}
```
i like how you refactored everything so your main function just "coordinates" the other classes!

denzelloke commented on src/main/java/Prune/Storage.java:
```diff
@@ -0,0 +1,82 @@
+            } catch (IOException e) {
+                System.out.println(e.getMessage());
+            }
+        }
```
the load function is a little long,
you could consider splitting it into the "happy path" - file exists
and the other path - save file doesnt exist --> make new file using a helper function

ashleyang2001 committed:
Reorganised README.md

ashleyang2001 committed:
Fix README improper format

ashleyang2001 committed:
Added Index out of bound exception for processMarking

ashleyang2001 committed:
Remove capitalization of package names.

ashleyang2001 committed:
Improved code quality, improved SLAP by extracting validation of storage file into a private method.

ashleyang2001 committed:
Final checks on coding standards.

ashleyang2001 committed:
Removed redundant file.

ashleyang2001 committed:
Fixed validateStorageFile(),  when storage file is not found, it creates relevant parent directories before creating the storage file.

ashleyang2001 committed:
Modified validateStorageFile() to remove redundant printing statements.

ashleyang2001 committed:
Added InvalidTaskFormatInFile exception to warn users of invalid task data format in storage file.

Added InvalidTaskFormatCommand exception specific to invalid task arguments for a given task type.
